File: src/python/problems/num_squares_test.py
#https://leetcode.com/problems/perfect-squares

from heapq import heappop, heappush
from core.problem_base import *

class NumSquares(ProblemBase):
    def Solution(self, n: int) -> int:
        map = [0] * (n + 1)

        for i in range(1, n + 1):
            map[i] = 2**31

            for j in range(1, i + 1):
                pow = j * j
                if i - pow < 0:
                    break

                map[i] = min(map[i], map[i - pow] + 1)

        return map[n]

if __name__ == '__main__':
    TestGen(NumSquares) \
        .Add(lambda tc: tc.Param("n", 1).Result(1)) \
        .Add(lambda tc: tc.Param("n", 13).Result(2)) \
        .Add(lambda tc: tc.Param("n", 4).Result(1)) \
        .Add(lambda tc: tc.Param("n", 3).Result(3)) \
        .Add(lambda tc: tc.Param("n", 12).Result(3)) \
        .Run()

File: src/go/problems/num_squares_test.go
// https://leetcode.com/problems/perfect-squares

package problems

import (
	"testing"
	"math"
	"github.com/dobermanch/leetcode/core"
)

type NumSquares struct{}

func TestNumSquares(t *testing.T) {
	gen := core.TestSuite[NumSquares]{}
	gen.Add(func(tc *core.TestCase) {
		tc.Param(4).Result(1)
	}).Add(func(tc *core.TestCase) {
		tc.Param(12).Result(3)
	}).Add(func(tc *core.TestCase) {
		tc.Param(13).Result(2)
	}).Add(func(tc *core.TestCase) {
		tc.Param(3).Result(3)
	}).Run(t)
}

func (NumSquares) Solution(n int) int {
    result := make([]int, n + 1)

    for i := 1; i <= n; i++ {
        result[i] = math.MaxInt32

        for j := 1; j <= i; j++ {
            pow := j * j
            if i - pow < 0 {
                break;
            }

            if result[i] > result[i - pow] + 1 {
                result[i] = result[i - pow] + 1
            }
        }
    }

    return result[n]
}
